Refactored `swipe` and moved `messageEdit` to a separate function.

DeclineThyself opened the pull request:
I have extracted my refactor of `swipe` and and a few more changes into a separate PR to make review of https://github.com/SillyTavern/SillyTavern/pull/4573 easier.

Commit:
Refactored `swipe` and moved `messageEdit` to a separate function.
Also, a few more minor changes.

DeclineThyself commented:
@Wolfsblvt Please run codex.
I'm now less sure of my testing.

DeclineThyself commented:
/gemini summary

DeclineThyself commented:
/gemini review

Wolfsblvt commented:
> @Wolfsblvt Please run codex. I'm now less sure of my testing.

I think you should be able to run it yourself, just comment `@codex review`.
I am tired tho, can't review this today.

Wolfsblvt commented:
@codex review

DeclineThyself commented:
> I am tired tho, can't review this today.

No rush, thank you.

> I think you should be able to run it yourself, just comment @codex review.

I tried, it wanted me to make an account with a phone number. I may later.

Wolfsblvt commented:
codex, are you stupid?

DeclineThyself commented:
> For now, I can only help with PRs you've created.

Huh, that worked before here.
https://github.com/SillyTavern/SillyTavern/pull/4573#pullrequestreview-3290587129

DeclineThyself commented on public/script.js:
```diff
@@ -8469,364 +8553,389 @@ export async function createOrEditCharacter(e) {
+        //Keep the swipe buttons at the same height when scrolling is finished.
+        let chatBottom = chatElement.scrollTop() + chatElement.height();
+        let messageBottom = this_mes_div.offset().top + this_mes_div.height();
+        let difference = (chatBottom - messageBottom);
```
This currently works to position the swipe arrows at the same scroll height whenever possible.

Wolfsblvt commented:
It's stupid. Lemme try again.

Wolfsblvt commented:
@codex review

DeclineThyself commented:
> It's stupid. Lemme try again.

Thank you.

Is there anything else you would like separated from this PR? https://github.com/SillyTavern/SillyTavern/pull/4573

DeclineThyself commented on public/script.js:
```diff
@@ -8469,364 +8553,389 @@ export async function createOrEditCharacter(e) {
-    chat[chat.length - 1]['swipe_id']--;
+    let generation;
+
+    const mesId = Number($(this).closest('.mes').attr('mesid') ?? chat.indexOf(message) ?? chat.length - 1);
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/eb83ccc1d62d50bdeeeca8ab66db616d98be9bf4
```javascript
const messageIndex = chat.indexOf(message);
if (messageIndex == -1 && typeof message != 'undefined') {
    console.error(`The message must exist in chat. ${message};`);
    return;
}
```

Commit:
Fixed bug.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2404789035
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2404789038

DeclineThyself commented on public/script.js:
```diff
@@ -8469,364 +8553,389 @@ export async function createOrEditCharacter(e) {
+        const swiped_messages_div  = chatElement.children().filter((index, div) => {
+            const $div = $(div);
+            return Number($div.attr('mesid')) >= mesId;
         });
```
Added a limit.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/eb83ccc1d62d50bdeeeca8ab66db616d98be9bf4
```javascript
//Select 20 messages after mesId.
const swiped_messages_div  = chatElement.children().filter((index, div) => {
    const $div = $(div);
    return mesId <= Number($div.attr('mesid')) && Number($div.attr('mesid')) <= mesId + 20;
});
```

Cohee1207 commented on public/script.js:
```diff
@@ -7872,18 +7931,43 @@ export function callPopup(text, type, inputValue = '', { okButton, rows, wide, w
+/**
+ * Swipe buttons are often toggled to update their position.
+ * This should be replaced with a more efficient function.
+ * This typically does not need to be awaited.
```
Remove async keyword from function definition, there's nothing to await

DeclineThyself replied:
Fixed, thank you.

Cohee1207 commented on public/script.js:
```diff
@@ -7893,35 +7977,34 @@ export function showSwipeButtons() {
-        swipeCounter.css('opacity', '0.3');
-    }
-    if ((chat[chat.length - 1].swipes.length - swipeId) === 1) {
+        swipeCounter.css('opacity', '0.3');    }
```
Please fix formatting on this line

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -7893,35 +7977,34 @@ export function showSwipeButtons() {
-    const swipeCounterText = formatSwipeCounter((swipeId + 1), chat[chat.length - 1].swipes.length);
+    const currentMessage = chatElement.children().filter(`[mesid="${mesId}"]`);
+    const swipeId = chat[mesId].swipe_id;
+    const swipeCounterText = formatSwipeCounter((swipeId + 1), chat[mesId]?.['swipes']?.length);
```
Optional chaining is not needed on this line

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -360,6 +361,7 @@ let default_user_name = 'User';
 export let name1 = default_user_name;
 export let name2 = systemUserName;
 export let chat = [];
+export let allow_swiping = true; //false when a swipe is in progress, or swiping is blocked.
```
It's better to use `isFooBar` for boolean naming, so `isSwipingAllowed`

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -5976,7 +5976,7 @@ export function syncSwipeToMes(messageId = null, swipeId = null) {
 
     const targetSwipeId = targetMessage.swipe_id;
-    if (!targetMessage.swipes[targetSwipeId] || !targetMessage.swipe_info[targetSwipeId]) {
+    if (typeof targetMessage.swipes?.[targetSwipeId] !== 'string' || typeof targetMessage.swipe_info?.[targetSwipeId] !== 'object') {
```
You don't need optional chaining, both are checked to be arrays above. Type checks are fine, I think.

DeclineThyself replied:
If I remember correctly, `swipe_info` can be empty while the swipe message is set.
This triggered and did not sync the swipe to the message.

Cohee1207 replied:
See the code directly above. Both arrays are checked to be arrays.

DeclineThyself replied:
You're correct, thank you.
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -7189,6 +7188,64 @@ function messageEditAuto(div) {
+
+    this_edit_mes_id = edit_mes_id;
+
+    let this_mes_div = chatElement.children().filter(`.mes[mesid="${edit_mes_id}"]`);
```
Since we're touching this legacy code, let's use `const` and update local variable naming to use camelCase

DeclineThyself replied:
Done.

Cohee1207 commented on public/script.js:
```diff
@@ -7221,9 +7278,11 @@ async function messageEditDone(div) {
         reasoningEditDone.trigger('click');
     }
 
+
```
Remove an added line

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -7221,9 +7278,11 @@ async function messageEditDone(div) {
     await eventSource.emit(event_types.MESSAGE_UPDATED, this_edit_mes_id);
     this_edit_mes_id = undefined;
     await saveChatConditional();
+    showSwipeButtons();
```
How did message edit restore swipe visibility previously?

DeclineThyself replied:
Currently in staging, swipe buttons are not hidden when messages are edited.
Instead, the message editor was closed and changes were discarded.
```javascript
  if (chat.length - 1 === Number(this_edit_mes_id)) {
      closeMessageEditor();
  }
  ```
  In refactor/swipe, `messageEdit` hides the swipe buttons.
  ```javascript
  async function messageEdit(edit_mes_id) {
    hideSwipeButtons();
  ```

Cohee1207 commented on public/scripts/RossAscends-mods.js:
```diff
@@ -1109,8 +1110,9 @@ export function initRossMods() {
+                allow_swiping &&
                 !isNanogallery2LightboxActive() &&  // Check if lightbox is NOT active
-                $('.swipe_left:last').css('display') === 'flex' &&
+                // $('.swipe_left:last').css('display') === 'flex' && //Redundant due to allow_swiping.
```
Remove the commented code

DeclineThyself replied:
Done.

Cohee1207 commented on public/scripts/RossAscends-mods.js:
```diff
@@ -1123,8 +1125,9 @@ export function initRossMods() {
+                allow_swiping &&
                 !isNanogallery2LightboxActive() &&  // Check if lightbox is NOT active
-                $('.swipe_right:last').css('display') === 'flex' &&
+                // $('.swipe_right:last').css('display') === 'flex' && //Redundant due to allow_swiping.
```
Remove the commented code

DeclineThyself replied:
Done.

Cohee1207 commented on public/style.css:
```diff
@@ -847,7 +847,7 @@ body .panelControlBar {
     padding-bottom: 0;
     overflow-y: scroll;
-    display: flex;
+    /* display: flex; This breaks swiping transitions when the message grows past chat's height. */
```
Though I suspect this change breaks a lot more things that it fixes. How did transitions work previously? Why did it stop now?

DeclineThyself replied:
Transitions were disabled here: https://github.com/SillyTavern/SillyTavern/blame/staging/public/script.js#L8565C50-L8565C50

This shouldn't break anything. The sheld is set to `display: flex;` https://github.com/SillyTavern/SillyTavern/pull/4610/files/eb83ccc1d62d50bdeeeca8ab66db616d98be9bf4#diff-46467272678c3da621692b13b96d47ed25128fa8027d8dbedd84684c6336fb51R757

Cohee1207 replied:
I'm afraid of further downstream impact of this change. `#chat` has at least several style that only work for display: flex, such as `flex-grow: 1` and `flex-direction: column`, this may also break third-party themes. Let's find an alternate approach to fix transitions without changing the display value.

DeclineThyself replied:
Thank you for explaining.
May I temporarily disable `display: flex;` in `expandNewMessage`?

Cohee1207 replied:
Can you demonstrate how exactly the transition breaks on flex?

DeclineThyself replied:
Yes.

https://github.com/user-attachments/assets/b13f000e-2524-4900-acf7-301b024d0b9f

Cohee1207 replied:
Why is it sliding around like this?

DeclineThyself replied:
During the recording, I slowed the animation duration to make the difference easily visible.

The sliding was the original behavior from Tavern before it was disabled here:
https://github.com/SillyTavern/SillyTavern/blame/41c048ed353138dbc1ce329ede4e6dcbe8b07147/public/script.js#L8704

Cohee1207 replied:
I think it's fine to have it disabled for good.

DeclineThyself replied:
> I think it's fine to have it disabled for good.

The sliding, or the `chat`'s `display: flex;`?

Cohee1207 replied:
The sliding.

DeclineThyself replied:
Done.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/49bd2d8b74d07e143c8235961f6aca2d6f4f5043

Cohee1207 replied:
Yet, the display is still commented out.

DeclineThyself replied:
I forgot, thank you.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/005485055b7b98840952fdce36854d7067700ac3

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+    return `${!isNaN(current) ? current : '?'}\u200b/\u200b${!isNaN(total) ? total : '?'}`;
 }
 
+
```
Please remove an added empty line

DeclineThyself replied:
Done.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
- * Handles the swipe to the left event.
+ * Handles the swipe event.
  * @param {JQuery.Event} _event Event.
+ * @param {boolean} swipe_right The direction to swipe.
```
It's better to make this argument a enum/named constant, e.g.

```js
export const SWIPE_DIRECTIONS = {
  LEFT: 'left',
  RIGHT: 'right',
};
```

DeclineThyself replied:
Done.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+export async function swipe(_event, swipe_right, { source, repeated, message = chat[chat.length - 1] } = {}) {
+
+    if (chat.length === 0) {
+        console.error('Swipe was called on an empty chat.');
```
Downgrade this to a warning level, it's not really an error

DeclineThyself replied:
Done.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
-    const swipe_duration = 120;
-    const swipe_range = '700px';
-    chat[chat.length - 1]['swipe_id']--;
+    const mesId = Number($(this).closest('.mes').attr('mesid') ?? messageIndex ?? chat.length - 1);
```
`messageIndex` can't be null/undefined, it's always >= -1

DeclineThyself replied:
Thank you, Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
-    // If the user is holding down the key and we're at the first swipe, don't do anything
-    if (source === 'keyboard' && repeated && chat[chat.length - 1].swipe_id === 0) {
+    const messageIndex = chat.indexOf(message);
+    if (messageIndex == -1 && typeof message != 'undefined') {
```
This logic breaks if you pass message as `null`. Check just the truthiness of `message`, e.g. `if (message && messageIndex === -1)`

DeclineThyself replied:
Thank you, Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+    async function swipeGenerate() {
+
+        //Cancel the generation if it's a user message.
+        if (chat[mesId].is_user || mesId === 0) {
```
This breaks generation on overswipe right on first message in tainted chats. It should allow swipe to generate if `chat_metadata['tainted'] == true`

DeclineThyself replied:
Fixed.
What is this meant for?

Cohee1207 replied:
Generating swipes on first messages.

DeclineThyself replied:
Thank you.

DeclineThyself replied:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R8727

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
-            },
+            let run_generate = true;
+            await syncWithSwipeId(mesId, run_generate);
+            await animateSwipe(run_generate = true);
```
This assignment looks like a typo

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+        }
+    }
+
+    //Deepseek-V3.1
```
Ok?..

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+     * @param {number} duration
+     */
+    async function animateSwipeTransition(mesId, x, duration) {
+        //Select 20 messages after mesId.
```
What is going on here?

DeclineThyself replied:
In chatTree all messages under mesId must be visually swiped.
Gemini wanted a limit, so I decided twenty messages is enough for most screen sizes.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2405213346

Cohee1207 replied:
There's no chat tree in this branch.

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
-                            easing: animation_easing,
-                            queue: false,
-                            complete: function () {
+        } ); // this moves the div back and forth
```
What is this comment.. No, it only moves it one direction, lol.

DeclineThyself replied:
There were many comments from the original code that I did not remove or correct.
I'll remove them.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+                this_mes_div.css('height', 'auto');
+                this_mes_div.css('overflow', '');
+                //Correct height auto offset.
+                this_mes_div[0].scrollIntoView(false);
```
Do not use scrollIntoView, it breaks layout on phones.

DeclineThyself replied:
Is this still true?
https://caniuse.com/?search=scrollIntoView
https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView#browser_compatibility

Cohee1207 replied:
Yes. The availability of function is not an issue. The function itself is.

DeclineThyself replied:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/9b922a69bd8131af6321e0e9fcfbf9854c65e54f

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
 
-    // Make sure ad-hoc changes to extras are saved before swiping away
-    syncMesToSwipe();
+    //Hide swipe buttons.
```
Please remove a pointless comment

DeclineThyself replied:
Removed.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
+        return;
+    }
+
+
```
Too many blank lines.

DeclineThyself replied:
Fixed.

Cohee1207 commented on public/script.js:
```diff
@@ -8469,364 +8553,395 @@ export async function createOrEditCharacter(e) {
     }
+    //User feedback.
+    else {
+        //Shake
```
What is going on here?

DeclineThyself replied:
Due to performance when chatTree is enabled, I do not hide or show swipe buttons when `hideSwipeButtons` or `showSwipeButtons` is called.
Instead, this shows that the message cannot be swiped. This can only trigger when chatTree is enabled.

I would prefer to use an [intersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) to track which messages are visible on screen.
Then update their swipe counters and buttons when they become visible.
I did not want to make this PR any longer.

Cohee1207 replied:
There is no chat tree in this branch. This change is unrelated to the refactoring

DeclineThyself replied:
I removed it.

Cohee1207 reviewed (changes requested):
It appears to be still working on some very happy cases, but the code requires some cleanup.

DeclineThyself commented:
Thank you for the review.
I will correct everything soon.

Commit:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408682277
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408689706
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408690772
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408697066
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408705156
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408708088
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408725971
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408726241
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408740050
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408745918
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408753165
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408761262
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408764531
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408781694
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408784426
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408794672
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408802366
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408803433

Cohee1207 commented on public/script.js:
```diff
@@ -7189,6 +7188,64 @@ function messageEditAuto(div) {
+    updateEditArrowClasses();
+}
+
 async function messageEditDone(div) {
```
Would it be better to extract `messageEditCancel` from the click handler and place it alongside other edit-related functions?

DeclineThyself replied:
Done.

Commit:
Fixed:
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408702506

https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408805635

https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2410368443

https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408791120

Cohee1207 committed:
Fix formatting

Cohee1207 reviewed (changes requested):
Bug found: alternate greetings on the first message don't work, showing the same text for all swipes.

DeclineThyself commented:
> Bug found: alternate greetings on the first message don't work, showing the same text for all swipes.

Thank you. I did not test alternate greetings.

The greetings have messages without any swipe_info.
Causing this to return false:
```javascript
if (typeof targetMessage.swipes[targetSwipeId] !== 'string' || typeof targetMessage.swipe_info[targetSwipeId] !== 'object') {
```

How's this fix?

```javascript
    const targetSwipeId = targetMessage.swipe_id;
    if (typeof targetMessage.swipes[targetSwipeId] !== 'string') {
        console.warn(`[syncSwipeToMes] Invalid swipe ID: ${targetSwipeId}`);
        return false;
    }

    const targetSwipeInfo = targetMessage?.swipe_info[targetSwipeId];
    if (typeof targetSwipeInfo !== 'object') {
        console.warn(`[syncSwipeToMes] Invalid swipe info: ${targetSwipeId}`);
    }

    targetMessage.mes = targetMessage.swipes[targetSwipeId];
    targetMessage.send_date = targetSwipeInfo?.send_date;
    targetMessage.gen_started = targetSwipeInfo?.gen_started;
    targetMessage.gen_finished = targetSwipeInfo?.gen_finished;
    targetMessage.extra = structuredClone(targetSwipeInfo?.extra) ?? {};

    return true;
}
```

This handles all edge cases I can think of.

Wolfsblvt commented:
We already have `syncMesToSwipe` and `syncSwipeToMes` that were made to make this easy and not have to do all the redundant checks all the time, or writing to the message.
Would it make sense for you to use them?

https://github.com/SillyTavern/SillyTavern/blob/bd2db7c4fd8a598578adfe7fe9aca037759eea45/public/script.js#L5887-L5998

Wolfsblvt commented on public/script.js:
```diff
@@ -8472,366 +8588,383 @@ export async function createOrEditCharacter(e) {
-    const swipe_duration = 120;
-    const swipe_range = '700px';
-    chat[chat.length - 1]['swipe_id']--;
+    const mesId = Number($(this).closest('.mes').attr('mesid') ?? messageIndex ?? chat.length - 1);
```
If you calculate the `mesId` above by checking the index in the chat array, do you need to find it in the DOM too again?

DeclineThyself replied:
To keep complete compatibility when extensions call `swipe_left` or `swipe_right` I do not assume that `message` will be set and valid.

Wolfsblvt commented on public/script.js:
```diff
@@ -8472,366 +8588,383 @@ export async function createOrEditCharacter(e) {
+    // if (isNaN(current) || isNaN(total)) {
+    //     return '';
+    // }
+    return `${!isNaN(current) ? current : '?'}\u200b/\u200b${!isNaN(total) ? total : '?'}`;
```
I think the swipe counter was hidden on purpose for those messages. It looks weird to have `?/`` there I feel.

DeclineThyself replied:
Sometimes either `swipe_id` or `swipes` is `undefined`, resulting in `?/3` or `3/?`.
Which I prefer, especially for debugging.

Should I revert it so if one is `undefined` it returns `' '`?  Or change it to ` /3` and `3/ ` if one is undefined?

I would prefer to use an [intersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) to track which messages are visible on screen.
Then update their swipe counters and buttons when they become visible.
I did not want to make this PR any longer.

DeclineThyself replied:
How's this @Wolfsblvt?
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/29b669c80c86c0dc608236af220a9bbddd4a187b

```javascript
function formatSwipeCounter(current, total) {
    if (isNaN(current) && isNaN(total)) {
        return '';
    }
    return `${!isNaN(current) ? current : '?'}\u200b/\u200b${!isNaN(total) ? total : '?'}`;
}
```

Wolfsblvt replied:
I would be okay with that. That's a good idea. To visibly show there should be swipes but something is broken.
But generally, messages *should* not be in that state. Then some extension or code did something wrong.

Wolfsblvt commented on public/script.js:
```diff
@@ -8472,366 +8588,383 @@ export async function createOrEditCharacter(e) {
+    const isPristine = !chat_metadata?.tainted;
+
+    const swipeDuration = 120;
+    let swipeRange = 700;
```
We have animation duration set up already that respects user settings.
It's also available in code, if you need it outside of CSS. There is a setter, and we can likely make a small getter, if needed.
https://github.com/SillyTavern/SillyTavern/blob/bd2db7c4fd8a598578adfe7fe9aca037759eea45/public/script.js#L732-L740

DeclineThyself replied:
I can use that instead.

`swipeDuration` was copied from `swipe_left` and `swipe_right`'s `swipe_duration`.
Cohee asked me to change the variables to camelCase.

https://github.com/SillyTavern/SillyTavern/blob/staging/public/script.js#L8542
```javascript
 $(this).parent().children('.mes_block').transition({
            x: swipe_range,
            duration: animation_duration > 0 ? swipe_duration : 0,
            easing: animation_easing,
            queue: false,
```

Do you want the original behavior changed to only use `animation_duration`?

Wolfsblvt replied:
Hmmh. I see. There might've been a reason why it was set to a fixed speed when animation was enabled.
I think the old behavior is fine. Or if you want your implementation, make sure to at least respect when `animation_duration` is set to 0.

Cohee1207 replied:
`animation_duration` can't go negative. A correct condition would be `> 0`

DeclineThyself replied:
How's this?
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/331117931e12d6b52c180d220659f6d74351fbd4

Cohee1207 replied:
swipeDuration makes no sense. Just use animation_duration.

Cohee1207 replied:
I adjusted the speed to a sensible value.

Wolfsblvt commented on public/script.js:
```diff
@@ -8472,366 +8588,383 @@ export async function createOrEditCharacter(e) {
+            addOneMessage(chat[mesId], { type: 'swipe', forceId: mesId, scroll: false });
+        }
+
+        //Fallback.
+        await updateSwipeCounter(mesId);
+        if (mesId != chat.length - 1) {
+            await updateSwipeCounter(chat.length - 1);
+        }
```
Why manually updating, doesn't `showSwipeButtons` already update/show the correct swipes counter too?

DeclineThyself replied:
I forgot to remove that from this branch.

`showSwipeButtons` is disabled in chatTree, this was a fallback.

Wolfsblvt reviewed (changes requested):
Okay. I'm sorry, I tried, but just by looking at the code my eyes got tired and I lost any kind of focus.
I know you might feel similar, but I fear I can't review this. Can't speak for Cohee, but..

You already split it into a different branch. But maybe actually see that this is split per feature either?
You are mixing a new animation/transition with allowing to show/hide swipes on older messages, a messageEdit (?) and the prep for the new generation on older messages.

Suggestion:
1. Split the animation changes off. That should be its own thing.
2. Add new functionality that allows showing/hiding swipe buttons on older messages. That should be simple refactor and **not** actually change the logic in any way. We can make it an internal exported toggle or something, to test. But so we don't mix actual logic into it too.
  I mean, older messages can already contain swipes. I know there are extensions who simply show them.
  They don't need to be able to trigger any new generations.
  Make sure slash commands work and are updated to respect the new feature flag
3. Add more of the functionality.

Step by step.
So we actually can see and trying to comprehend what you are trying to do. And whatever AI tool you are using, should also stay on a single focus, and not change several dozen unrelated things.

DeclineThyself commented:
> We already have `syncMesToSwipe` and `syncSwipeToMes` that were made to make this easy and not have to do all the redundant checks all the time, or writing to the message. Would it make sense for you to use them?
> 
> https://github.com/SillyTavern/SillyTavern/blob/bd2db7c4fd8a598578adfe7fe9aca037759eea45/public/script.js#L5887-L5998

syncSwipeToMes is breaking. I'm am proposing a to these lines:
https://github.com/SillyTavern/SillyTavern/blob/bd2db7c4fd8a598578adfe7fe9aca037759eea45/public/script.js#L5918-L5935

DeclineThyself commented:
> So we actually can see and trying to comprehend what you are trying to do. And whatever AI tool you are using, should also stay on a single focus, and not change several dozen unrelated things.

I copied code from an LLM once, which was clearly labeled.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408787781

The rest was written by me and those that came before. I apologize for my many mistakes.

Wolfsblvt commented:
> syncSwipeToMes is breaking. 

What exactly is breaking? Maybe we can fix that first?
I has like eight if checks or something checking the state is somewhat consistent before so it doesn't break anything.

Wolfsblvt commented:
> The rest was written by me and those that came before. I apologize for my many mistakes.

You don't have to apologize, it's all good. We all make mistakes.
I am just trying to point out that this will be very hard to test with so many things mixed together. And you can see by the many pointers at code that new changes might introduce new issues.

That's why I am seeing if we can try if we can split it into other chunks that do not touch code so far spread.

Commit:
Bugfixes. Incomplete.

DeclineThyself commented:
> > syncSwipeToMes is breaking.
> 
> What exactly is breaking? Maybe we can fix that first? I has like eight if checks or something checking the state is somewhat consistent before so it doesn't break anything.

I didn't commit because I wasn't finished.

Here is the problem:
https://github.com/SillyTavern/SillyTavern/pull/4610#issuecomment-3378976066

Here is my proposed fix to `syncSwipeToMes`:
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/8fa79c16df5de8c29026d8852beecf816d4bbb8d#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R5976-R5993

Wolfsblvt commented:
Ah I see. `syncSwipeToMes` and `syncMesToSwipe` are both intended to only work and do their stuff if the messages are correctly set up for "swipes", meaning they have the info object, etc. So we don't add swipe data before cloning in other functions, or to messages which shouldn't have swipes.

Generally, we could introduce an `{ setup = false }` argument or something similar that allows to call the utility and also fix the message/swipe.
But honestly, that is weird. If the message had no swipe set up, we should not write swipe data to the message. And the other way around too.
Likely a separate "prepareForSwipeData" or "ensureSwipesData" or something is needed.

DeclineThyself commented:
> Ah I see. `syncSwipeToMes` and `syncMesToSwipe` are both intended to only work and do their stuff if the messages are correctly set up for "swipes", meaning they have the info object, etc. So we don't add swipe data before cloning in other functions, or to messages which shouldn't have swipes.

The greeting should be swipeable, but it never has swipe_info.
Should I alter `swipe` to always set `chat[mesId]['swipe_info'][targetSwipeId] ??= {}`?

My current proposed fix currently sets `send_date`, `gen_started` and `gen_finished` to `undefined` and `extra` to `{}` if they do not exist.

Instead, would this work?

```javascript

    targetMessage.mes = targetMessage.swipes[targetSwipeId];
    targetMessage.send_date = targetSwipeInfo?.send_date ?? targetMessage.send_date;
    targetMessage.gen_started = targetSwipeInfo?.gen_started ?? targetMessage.gen_started;
    targetMessage.gen_finished = targetSwipeInfo?.gen_finished ?? targetMessage.gen_finished;
    targetMessage.extra = structuredClone(targetSwipeInfo?.extra) ?? targetMessage.extra;
```

Should this swap the `targetMessage` to whatever is valid or leave it unchanged?

DeclineThyself commented:
> Okay. I'm sorry, I tried, but just by looking at the code my eyes got tired and I lost any kind of focus. I know you might feel similar, but I fear I can't review this. Can't speak for Cohee, but..

Please don't apologize. These have been my first PRs and I have made many mistakes. Thank you for your kind guidance.
I am also exhausted with the `swipe_left` and `swipe_right` refactor.

> Suggestion:
> 
>     1. Split the animation changes off. That should be its own thing.

I only modified this animation:
https://github.com/SillyTavern/SillyTavern/pull/4610/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L8762-L8774
The duration was set to zero, so I attempted to fix it here:
 https://github.com/SillyTavern/SillyTavern/pull/4610/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R8768-R8785
 Should I separate it from this PR?
 
>     2. Add new functionality that allows showing/hiding swipe buttons on older messages. That should be simple refactor and **not** actually change the logic in any way. We can make it an internal exported toggle or something, to test. But so we don't mix actual logic into it too.
>        I mean, older messages can already contain swipes. I know there are extensions who simply show them.
>        They don't need to be able to trigger any new generations.
>        Make sure slash commands work and are updated to respect the new feature flag

Showing/Hiding swipe buttons is not contained in this PR, Should I separate that from the chatTree PR?

I completely forgot about slash commands.
Should I include that in the chatTree PR Or keep in in a separate branch?

>     3. Add more of the functionality.

What functionality should be removed from this PR?

>I am just trying to point out that this will be very hard to test with so many things mixed together. And you can see by the many pointers at code that new changes might introduce new issues.
That's why I am seeing if we can try if we can split it into other chunks that do not touch code so far spread.

I can try. I have been avoiding separate branches because I am dreading the merge conflicts.
Should I focus on splitting this PR or https://github.com/SillyTavern/SillyTavern/pull/4573?

Commit:
Use scrollTop because scrollIntoView breaks layout on phones.

Commit:
Only show '?' in `formaSwipeCounter` if something is wrong.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2412169005

Cohee1207 commented:
> I can try. I have been avoiding separate branches because I am dreading the merge conflicts.
> Should I focus on splitting this PR or https://github.com/SillyTavern/SillyTavern/pull/4573?

Merge conflicts are unavoidable in a live codebase. Just focus on one thing at a time. There's absolutely no rush to deliver everything in one release. We can do it step by step... Let's probably start with message editor refactoring if that's okay with you

DeclineThyself commented:
> > I can try. I have been avoiding separate branches because I am dreading the merge conflicts.
> > Should I focus on splitting this PR or #4573?
> 
> Merge conflicts are unavoidable in a live codebase. Just focus on one thing at a time. There's absolutely no rush to deliver everything in one release. We can do it step by step... Let's probably start with message editor refactoring if that's okay with you

Ok, I'll split that into a new PR now.
Thank you again for your patience and guidance.

Commit:
Extracted `messageEdit` and `messageEditCancel` from `.mes_edit` and `.mes_edit_cancel`

Commit:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4633#discussion_r241505
https://github.com/SillyTavern/SillyTavern/pull/4633#pullrequestreview-3316588180

DeclineThyself commented:
Rebased on [refactor/message-edit](https://github.com/SillyTavern/SillyTavern/pull/4633).

DeclineThyself commented:
Rebased on https://github.com/SillyTavern/SillyTavern/pull/4633.

DeclineThyself commented:
> Merge conflicts are unavoidable in a live codebase. Just focus on one thing at a time. There's absolutely no rush to deliver everything in one release. We can do it step by step... Let's probably start with message editor refactoring if that's okay with you

@Cohee1207, What else should I refactor out of this PR?
`refreshSwipeButtons` and the spelling errors?

Commit:
Respect animation_duration if it's >= 0.

Commit:
Disabled expandNewMessage's animation on Cohee's request. https://github.com/SillyTavern/SillyTavern/pull/4610/files#r2408731744

DeclineThyself commented on public/script.js:
```diff
@@ -8074,6 +8097,7 @@ export function showSwipeButtons() {
 export function hideSwipeButtons() {
+    isSwipingAllowed = false;
     chatElement.find('.swipe_right').hide();
     chatElement.find('.last_mes .swipes-counter').hide();
```
```suggestion
```

https://github.com/SillyTavern/SillyTavern/pull/4666#issuecomment-3415955322
> Also since now swipes counters are hidden on message edit, you won't be able to see if the message has any swipes to delete or not.

@Cohee1207 May I change this to never hide the `.swipes-counter`?

Cohee1207 replied:
I'm not sure. Including when the generation is ongoing?

DeclineThyself replied:
Yes, I would prefer to.
Or, I can move it to a separate `hideSwipesCounter` function.

Cohee1207 replied:
Maybe as a parameter to hideSwipeButtons? Default value: true.
Edit could call it like:
`hideSwipeButtons({ hideCounters: false })`

DeclineThyself replied:
Why do you prefer one function that does multiple things instead of explicit functions?

I was taught differently:
https://stackoverflow.com/questions/18045574/is-it-better-practice-to-have-2-functions-or-1-complex-function
https://softwareengineering.stackexchange.com/a/391162
https://stackoverflow.com/questions/12994628/c-design-single-function-performing-multiple-tasks-vs-multiple-functions-perf

DeclineThyself replied:
> I'm not sure. Including when the generation is ongoing?

Does the `swipes-counter` only need to be hidden when generating?

Cohee1207 replied:
Because there's already a ton of functions in script.js and I doubt there's any value in showing/hiding labels separately from the buttons.

Cohee1207 replied:
> Does the swipes-counter only need to be hidden when generating?

I'm confident that counters need to be shown on edit. The rest - we'll see. I haven't seen it all or tried it in action yet.

DeclineThyself replied:
> Because there's already a ton of functions in script.js and I doubt there's any value in showing/hiding labels separately from the buttons.

Thank you for explaining.

DeclineThyself replied:
> Maybe as a parameter to hideSwipeButtons? Default value: true. Edit could call it like: `hideSwipeButtons({ hideCounters: false })`

https://github.com/SillyTavern/SillyTavern/pull/4610/commits/c777a88b3425cef5c9db703babd1cda3aa68a64c
The counter is now hidden while a swipe is generating.
When should counter be hidden?
Should it be hidden for old messages?
I'll change the default to hidden once I know when it can be visible.

I would prefer to always leave it visible, I may be in the minority.

DeclineThyself replied:
@Cohee1207 

> The counter is now hidden while a swipe is generating.
When should counter be hidden?
Should it be hidden for old messages?

Commit:
Only hide the swipe counter when a generation is ongoing.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2440588779

Commit:
refreshSwipeButtons

Cohee1207 committed:
Adjust swipeDuration constant

Cohee1207 committed:
feat: add refreshSwipeButtons and isSwipingAllowed to context

DeclineThyself commented on public/script.js:
```diff
@@ -2920,7 +2920,7 @@ class StreamingProcessor {
             this.markUIGenStarted();
         }
-        hideSwipeButtons({ hideCounters:true });
+        hideSwipeButtons({ hideCounters: true });
```
Thank you. I'll check my spacing in the future.

Cohee1207 commented on public/script.js:
```diff
@@ -8677,366 +8706,381 @@ export async function createOrEditCharacter(e) {
+            chat[mesId]['swipe_id'] = originalSwipeId;
+            //Prepare to generate.
+        } else {
+            if (chat[mesId].extra) {
```
I think this needs to run on all swipes, including swipe left, otherwise image swipes will get stuck if you are generating them on a second swipe.

DeclineThyself replied:
Thank you, fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/5a8e1ab562de3f0f2b28817e027606d1cde47eae

DeclineThyself replied:
@Wolfsblvt, I believe `syncSwipeToMes` is your function. May I alter this line to warn without `return false`?

```javascript
if (typeof targetSwipeInfo !== 'object') {
    return false;
    console.warn(`[syncSwipeToMes] Invalid swipe info: ${targetSwipeId}`);
}
  ```
  If changed,`syncSwipeToMes` would update the `swipe_info` on all messages, including swipe generations.
  ```javascript
targetMessage.mes = targetMessage.swipes[targetSwipeId];
targetMessage.send_date = targetSwipeInfo?.send_date;
targetMessage.gen_started = targetSwipeInfo?.gen_started;
targetMessage.gen_finished = targetSwipeInfo?.gen_finished;
targetMessage.extra = structuredClone(targetSwipeInfo?.extra) ?? {};
```
    
Then I could delete this.
https://github.com/SillyTavern/SillyTavern/pull/4610/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R8819-R8827
```javascript
if (chat[mesId].extra) {
    // if message has memory attached - remove it to allow regen
    delete chat[mesId].extra.memory;

    // ditto for display text
    delete chat[mesId].extra.display_text;

    delete chat[mesId].extra.image;
    delete chat[mesId].extra.image_swipes;
    delete chat[mesId].extra.video;
    delete chat[mesId].extra.inline_image;
}
delete chat[mesId].gen_started;
delete chat[mesId].gen_finished;
```

Cohee1207 commented on public/script.js:
```diff
@@ -8677,366 +8706,381 @@ export async function createOrEditCharacter(e) {
+                    delete chat[mesId].extra.memory;
+                }
+                // ditto for display text
+                if (chat[mesId].extra.display_text) {
```
It will be fine to run `delete chat[mesId].extra.display_text` and `delete chat[mesId].extra.memory` without additional checks

DeclineThyself replied:
Fixed. https://github.com/SillyTavern/SillyTavern/pull/4610/commits/5a8e1ab562de3f0f2b28817e027606d1cde47eae
I copied that from here.
https://github.com/SillyTavern/SillyTavern/blob/5eb43e45b609a7f7f92c54140398cc9d50cd4cc2/public/script.js#L8737-L8745

Cohee1207 commented:
The code is mostly fine with some open questions left. Running some quick tests, all good so far, but I would like to take more time for verification. If anyone would like to join testing, be my guest.

Cohee1207 commented:
Found a bug: swipe left when there's only one swipe on the message. Buttons are gone.

Commit:
Fixed image swipes.

https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2442489827

https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2442490042

Commit:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610#issuecomment-3418657967

DeclineThyself commented:
> Found a bug: swipe left when there's only one swipe on the message. Buttons are gone.

Fixed: https://github.com/SillyTavern/SillyTavern/pull/4610/commits/68096cadfb8204f0000a9d55b38ec154e3c522ca

When you asked me to remove the user feedback, I mistakenly removed `endSwipe`.

https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2408805635
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/876c5590cc74c2c377350c92bae9f5066beaa5e4#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L9054-L9062

DeclineThyself commented:
/gemini review

Commit:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2442480238

DeclineThyself commented on public/script.js:
```diff
@@ -6059,21 +6058,21 @@ export function syncSwipeToMes(messageId = null, swipeId = null) {
+    targetMessage.send_date = targetSwipeInfo?.send_date;
+    targetMessage.gen_started = targetSwipeInfo?.gen_started;
+    targetMessage.gen_finished = targetSwipeInfo?.gen_finished;
+    targetMessage.extra = structuredClone(targetSwipeInfo?.extra) ?? {};
```
Tested.
This does not crash, it correctly returns `{}`.
```javascript
targetSwipeInfo = null
structuredClone(targetSwipeInfo?.extra) ?? {};
```

Commit:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2442536011

DeclineThyself commented on public/script.js:
```diff
@@ -8677,364 +8706,373 @@ export async function createOrEditCharacter(e) {
+    const thisMesDiv = chatElement.children().filter(`.mes[mesid="${mesId}"]`);
+    const thisMesText = thisMesDiv.find('.mes_block .mes_text');
+    const thisMesDivHeight = thisMesDiv[0].scrollHeight;
+    const thisMesTextHeight = thisMesText[0].scrollHeight;
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4610/commits/667508a8aa4aa5c41ab9a9f4e74c7fd45714d7b6

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/script.js:
```diff
@@ -6059,21 +6058,21 @@ export function syncSwipeToMes(messageId = null, swipeId = null) {
-    const targetSwipeInfo = targetMessage.swipe_info[targetSwipeId];
+    const targetSwipeInfo = targetMessage?.swipe_info[targetSwipeId];
     if (typeof targetSwipeInfo !== 'object') {
-        return false;
+        console.warn(`[syncSwipeToMes] Invalid swipe info: ${targetSwipeId}`);
     }
```
See https://github.com/SillyTavern/SillyTavern/pull/4610#issuecomment-3378976066 and https://github.com/SillyTavern/SillyTavern/pull/4610#issuecomment-3379211586.

DeclineThyself commented:
/gemini summary

Commit:
Minor refactor.

Cohee1207 reviewed (approved):
I think I'm done with testing. But I'd like more input from people testing.

@RossAscends @Wolfsblvt I need your input

Cohee1207 commented on public/script.js:
```diff
@@ -8677,366 +8706,391 @@ export async function createOrEditCharacter(e) {
 
-    // Make sure ad-hoc changes to extras are saved before swiping away
-    syncMesToSwipe();
+    hideSwipeButtons();
```
Wishlist: hide the counters here, since streamed gen will hide them anyway, so they'll be visible for a split second until the stream starts, which is a bit jarring.

```suggestion
    hideSwipeButtons({ hideCounters: true });
```

DeclineThyself replied:
I would prefer leave this.
Please test this on an API that takes a few seconds for the first token.
When you swipe a message, the counter will swap from 1/1 to 2/1 as the message waits for generation to start.
Then it is hidden when the generation starts streaming.
Or if the generation is canceled, it switches back to 1/1.

Cohee1207 replied:
I also noticed a difference with staging where swipe buttons disappear for a split second when switching between swipes that don't require generation. So _ideally_ swipe function shouldn't call hide buttons, it should be done by the generation function as it is now.

Cohee1207 replied:
Removing this line fixes all the problems I have with swipe buttons flickering, except the misaligned swipes counter.
So if 2/1 is a problem, hideSwipes in generate should also hide labels.

Cohee1207 replied:
This patch solved it all for me. What do you think?
```diff
diff --git a/public/script.js b/public/script.js
index a2fb03c1f..74bb9dc48 100644
--- a/public/script.js
+++ b/public/script.js
@@ -3577,7 +3577,7 @@ export async function Generate(type, { automatic_trigger, force_name2, quiet_pro
         }
 
         // Hide swipes if not in a dry run.
-        hideSwipeButtons();
+        hideSwipeButtons({ hideCounters: true });
         // If generated any message, set the flag to indicate it can't be recreated again.
         chat_metadata['tainted'] = true;
     }
@@ -8079,12 +8079,17 @@ export function callPopup(text, type, inputValue = '', { okButton, rows, wide, w
 /**
  * Update the swipe counter for mesId.
  * @param {Number} mesId
+ * @param {object} [options] Options
+ * @param {boolean} [options.show=true] Whether to show the counter after updating.
  */
-export async function updateSwipeCounter(mesId) {
+export async function updateSwipeCounter(mesId, { show = true } = {}) {
     const swipeCounterText = formatSwipeCounter((chat[mesId]?.['swipe_id'] + 1), chat[mesId]?.['swipes']?.length);
     const currentMessage = chatElement.children().filter(`[mesid="${mesId}"]`);
     const swipeCounter = currentMessage.find('.swipes-counter');
-    swipeCounter.text(swipeCounterText).show();
+    swipeCounter.text(swipeCounterText);
+    if (show) {
+        swipeCounter.show();
+    }
 }
 
 /**
@@ -8909,7 +8913,8 @@ export async function swipe(_event, direction, { source, repeated, message = cha
         await animateSwipeTransition(mesId, swipeRange, swipeDuration);
 
         if (run_generate) {
-            await updateSwipeCounter(mesId);
+            hideSwipeButtons({ hideCounters: true });
+            await updateSwipeCounter(mesId, { show: false });
             //shows "..." while generating
             thisMesDiv.find('.mes_text').html('...');
             // resets the timer
@@ -8970,8 +8975,6 @@ export async function swipe(_event, direction, { source, repeated, message = cha
         return unblockGeneration();
     }
 
-    hideSwipeButtons();
-
     //If the swipe is not being deleted.
     if (source != 'delete') {
```

DeclineThyself replied:
>  So if 2/1 is a problem, hideSwipes in generate should also hide labels.

I prefer that 2/1 is shown to the user.

Is there any a scenario where swipe counters need to be updated in advance without being shown?
I do not understand when this could be needed.
```javascript
export async function updateSwipeCounter(mesId, { show = true } = {}) {
```

> it should be done by the generation function as it is now.

How's this?

https://github.com/SillyTavern/SillyTavern/pull/4610/commits/d9b046a0ea7697cdd4cb69296dfb4176845fafe5

DeclineThyself replied:
> How's this?
> 
> [d9b046a](https://github.com/SillyTavern/SillyTavern/commit/d9b046a0ea7697cdd4cb69296dfb4176845fafe5)

I originally used `hideSwipeButtons` in `swipe` to set `isSwipingAllowed` to false earlier.
Otherwise, holding the arrow key caused problems.
Setting ` isSwipingAllowed = false;` directly works.

Cohee1207 replied:
Click registration on buttons was a bit broken, I fixed it.

DeclineThyself replied:
> Click registration on buttons was a bit broken, I fixed it.

Thank you.

Cohee1207 commented on public/script.js:
```diff
@@ -8677,366 +8706,391 @@ export async function createOrEditCharacter(e) {
+ * @param {boolean} [params.repeated] Is the swipe event repeated.
+ * @param {object} [params.message] The chat message to swipe.
+ */
+export function swipe_left(_event, { source, repeated, message } = {}) {
```
Wishlist: since `swipe` is async, a caller of deprecated functions has no ability to tell when the swipe is completed. Mark this function as async and await the swipe.call promise. Same with `swipe_right`

DeclineThyself replied:
Fixed. https://github.com/SillyTavern/SillyTavern/pull/4610/commits/74133ef3eca7b426c076def67b6a76141d362bd5

Cohee1207 commented on public/script.js:
```diff
@@ -8677,366 +8706,391 @@ export async function createOrEditCharacter(e) {
 }
 
+/**
+ * DEPRECATED
```
Wishlist: there's a `@deprecated` JSDoc tag. Also it usually tells what to use instead.

DeclineThyself replied:
Fixed. https://github.com/SillyTavern/SillyTavern/pull/4610/commits/74133ef3eca7b426c076def67b6a76141d362bd5

Commit:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2443359660 https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2443359435

Commit:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4610#discussion_r2443357913

Cohee1207 committed:
Fix registration of click events

Cohee1207 committed:
Fix passing data to swipe events

Cohee1207 commented:
Well, I tried to get someone else to test this, but no one else is available at the moment. I can only pray that nothing breaks with this. Let's test in production...

DeclineThyself commented:
> Well, I tried to get someone else to test this, but no one else is available at the moment. I can only pray that nothing breaks with this. Let's test in production...

I do not mind waiting a few days for testers to be available, I would rather slow down than break things.

If clean up `swipe` more, should I commit to this branch in a new pull request?

Cohee1207 commented:
> If clean up swipe more, should I commit to this branch in a new pull request?

Create a new branch.

Cohee1207 commented:
> I do not mind waiting a few days for testers to be available, I would rather slow down than break things.

I'm afraid this isn't going to change any time soon. There's a general lack of interest of users in testing emerging feature branches.

DeclineThyself commented:
> I'm afraid this isn't going to change any time soon. There's a general lack of interest of users in testing emerging feature branches.

What do you think of adding a `tester` role to the discord with a colored name, and pinging the role when new features need to be tested?
A visible but locked `tester` channel would likely generate enough interest.

Cohee1207 commented:
> What do you think of adding a tester role to the discord with a colored name, and pinging the role when new features need to be tested?

It exists. But I don't want to ping 1700 people at once.

DeclineThyself commented:
> It exists. But I don't want to ping 1700 people at once.

I didn't know.
How about another role for people that don't mind being pinged and want to help with small changes?

underscorex86 commented:
beautiful beautiful lads! let's goooo

> I didn't know. How about another role for people that don't mind being pinged and want to help with small changes?

tell me what ya need tested!

DeclineThyself commented:
> tell me what ya need tested!

Thanks!
If you can find the time, please help me find bugs in https://github.com/SillyTavern/SillyTavern/pull/4573#issuecomment-3346598451
Are there any features or QOL improvements you would like?

```bash
git remote add swipes "https://github.com/DeclineThyself/SillyTavern"
git fetch swipes
git switch feat/chat-tree
```
To switch back to the `staging` branch, run `git switch staging`
